### ISSUE TITLE
AudioEffectCapture and AudioStreamGenerator producing slowed audio since after 4.4-dev3

### ISSUE THREAD

**Issue by Meorge** [godotengine/godot#99380]
### Tested versions

- Reproducible in v4.4.dev.custom_build [5efd124ca]
- Not reproducible in v4.4.dev3.official [f4af8201b]

### System information

Godot v4.4.dev (5efd124ca) - macOS 15.1.0 - Multi-window, 2 monitors - OpenGL 3 (Compatibility) - AMD Radeon Pro 5500 XT OpenGL Engine - Intel(R) Core(TM) i7-10700K CPU @ 3.80GHz (16 threads)

### Issue description

I have a setup where I route all of the output from a bus into an AudioEffectCapture, then populate an AudioStreamGenerator's buffer with its output. This allows me to play a bus's sound from a given AudioStreamPlayer2D/3D.
```gdscript
extends AudioStreamPlayer2D

@onready var playback: AudioStreamGeneratorPlayback
var capture_effect: AudioEffectCapture

func _ready():
	play()
	playback = get_stream_playback()
	var bus_index = AudioServer.get_bus_index("Capturer")
	var num_effects = AudioServer.get_bus_effect_count(bus_index)
	capture_effect = AudioServer.get_bus_effect(bus_index, num_effects - 1)

func _process(delta):
	var frames_available = capture_effect.get_frames_available()
	if playback.can_push_buffer(frames_available):
		var buf = capture_effect.get_buffer(frames_available)
		capture_effect.clear_buffer()
		playback.push_buffer(buf)
```
(I had implemented this in C++ with GDExtension for speed's sake, but ported it over to GDScript to confirm the issue persisted with it as well.)

In 4.4-dev3, this sounds pretty correct:

https://github.com/user-attachments/assets/1206ae3c-dace-4b71-8b68-0ad3d4294f3b

But in 4.4 on the master branch, it is slightly pitched down and crackles more:

https://github.com/user-attachments/assets/1e305e88-32cb-4a87-9e26-2d438f3105fd

### Steps to reproduce

MRPs are provided below for 4.4-dev3 and the master branch. They should be identical with the possible exception of structural project format changes made since dev3.

To test them, simply open and run `scene.tscn` (which is also the projects' main scene). When on 4.4-dev3 you should hear the music normally, and when on a version built from the master branch you should hear it slowed down, as seen in the videos above.

### Minimal reproduction project (MRP)

4.4-dev3 compatible project (where it sounds as intended): [lagging-audio-forwarding-4.4dev3.zip](https://github.com/user-attachments/files/17794352/lagging-audio-forwarding-4.4dev3.zip)

v4.4.dev.custom_build [5efd124ca] compatible project (where it sounds slowed and crackles): [lagging-audio-forwarding-4.4master.zip](https://github.com/user-attachments/files/17794359/lagging-audio-forwarding-4.4master.zip)

**Comment by Meorge** [comment]
Good news on this! I've determined this regression happened with PR #88628.

It looks like there, mix rate methods may have been implemented for some parts of the audio system, but not all (possibly excluding AudioEffectCapture or AudioStreamGenerator)? I'll try to look a bit more into it soon.

**Comment by Meorge** [comment]
Even better news! I think I understand why the problem occurs, and how to fix it.

After PR #88628, the output audio sampling rate is determined by CoreAudio. I can see when running Godot in verbose mode that, for me, it's 48,000 Hz. However, by default, the AudioStreamGenerator uses a mix rate of 44,100 Hz.

So, based on my understanding of audio, what I *think* is happening is that the Capture effect is outputting samples meant for a 48,000 Hz mix rate, but the Generator processes them at a 44,100 Hz mix rate. Since that's lower than what the Capture effect is outputting, the resulting audio sounds slowed and pitched down.

If I set the Generator's mix rate to 48,000 Hz, then the music sounds normal:

https://github.com/user-attachments/assets/1aac6f3f-99e8-4e95-aa2e-4bedf347fbbf

I'm not sure how exactly this would be "fixed", since it seems to be technically working correctly - it's just unclear to the user what is causing the problem. Perhaps the AudioStreamGenerator could have a boolean property that sets the mix rate to whatever the system mix rate is, rather than having the user manually set its value?

**Comment by bruvzg** [comment]
This looks like an issue with your script that is reading raw frames from one stream and writing it to another without resampling (or setting mix rates to match). But documentation probably should be updated to note how default `max_rate` is used. And maybe `AudioStreamGenerator` should have presets to automatically match either input or output rate of the `AudioServer` if it's used with a captured stream.

**Comment by Meorge** [comment]
Yep, that is my understanding as well. I modified my test script to pull the `mix_rate` from `AudioServer`, like so:
```gdscript
extends AudioStreamPlayer2D

@onready var playback: AudioStreamGeneratorPlayback
var capture_effect: AudioEffectCapture

func _ready():
	play()

        # Set mix rate to what the AudioServer is outputting
	(stream as AudioStreamGenerator).mix_rate = AudioServer.get_mix_rate()
	playback = get_stream_playback()
	var bus_index = AudioServer.get_bus_index("Capturer")
	var num_effects = AudioServer.get_bus_effect_count(bus_index)
	capture_effect = AudioServer.get_bus_effect(bus_index, num_effects - 1)

func _process(delta):
	var frames_available = capture_effect.get_frames_available()
	if playback.can_push_buffer(frames_available):
		var buf = capture_effect.get_buffer(frames_available)
		capture_effect.clear_buffer()
		playback.push_buffer(buf)
```
And with this, it works! 🥳 

What I would consider the bug behavior is that, as far as I can see in the project settings, there's no indication that the set "Mix Rate" is being overridden with a different value.

![Image](https://github.com/user-attachments/assets/296a40d6-736f-4eca-8cec-f6d626288c50)

Renaming this setting to something like "Target Mix Rate" could help with the confusion, but I imagine that'd also be a breaking change (I think the names in Project Settings are generated from the internal property names?)

Based on what I know of the Project Settings UI, this may be hard to do, but it could be nice for it to detect if the user is on macOS/iOS/iPadOS (or any platform affected by #88628), or if the project export settings target at least one of those platforms, and if so display an extra little warning message that the mix rate may be different on those platforms.

I definitely like the idea of `AudioStreamGenerator` having presets to just use whatever the `AudioServer`'s input/output rate is.
